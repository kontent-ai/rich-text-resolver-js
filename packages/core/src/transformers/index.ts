export * from "./html-transformer/index.js";
export * from "./portable-text-transformer/index.js";
export * from "./transformer-models.js";
