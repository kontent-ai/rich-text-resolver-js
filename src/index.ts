export { toVueImageDefault } from './utils/resolution/vue.js';
export { toHTMLImageDefault } from './utils/resolution/html.js';
export { traversePortableText } from './utils/transformer-utils.js';
export { transformToJson } from "./transformers/json-transformer/json-transformer.js";
export { transformToPortableText } from "./transformers/portable-text-transformer/portable-text-transformer.js";
export { parse as browserParse } from "./parser/browser/rich-text-browser-parser.js";
export { parse as nodeParse } from "./parser/node/rich-text-node-parser.js";
