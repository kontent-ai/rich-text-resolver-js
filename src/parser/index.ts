import * as runtimeEnvironment from "browser-or-node";
import { browserParse } from "./browser/index.js";
import { nodeParse } from "./node/index.js";

export * from "./parser-models.js";

export const parseHTML = (htmlInput: string) =>
  runtimeEnvironment.isBrowser
    ? browserParse(htmlInput)
    : nodeParse(htmlInput);
