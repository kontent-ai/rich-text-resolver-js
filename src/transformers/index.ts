export * from "./json-transformer/index.js";
export * from "./portable-text-transformer/index.js";
export * from "./transformer-models.js";
