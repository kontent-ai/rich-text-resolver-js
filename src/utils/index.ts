export * from "./browser-parser-utils.js";
export * from "./common-utils.js";
export * from "./constants.js";
export * from "./transformer-utils.js";
