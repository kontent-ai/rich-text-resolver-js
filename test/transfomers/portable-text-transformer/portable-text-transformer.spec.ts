import { Elements, ElementType } from "@kontent-ai/delivery-sdk";
import { escapeHTML, PortableTextOptions, toHTML } from '@portabletext/to-html';
import { nodeParse, resolveTable, transform } from "../../../src";

jest.mock('crypto', () => {
  return {
    randomUUID: jest.fn(() => 'guid')
  }
});

const dummyRichText: Elements.RichTextElement = {
  value: "<p><br></p>",
  type: ElementType.RichText,
  images: [],
  linkedItemCodenames: [],
  linkedItems: [
    {
      system: {
        id: "99e17fe7-a215-400d-813a-dc3608ee0294",
        name: "test item",
        codename: "test_item",
        language: "default",
        type: "test",
        collection: "default",
        sitemapLocations: [],
        lastModified: "2022-10-11T11:27:25.4033512Z",
        workflowStep: "published"
      },
      elements: {
        text_element: {
          type: ElementType.Text,
          name: "text element",
          value: "random text value"
        }
      }
    }
  ],
  links: [],
  name: "dummy"
};

const portableTextComponents: PortableTextOptions = {
  components: {
    types: {
      image: ({value}) => {
        return `<img src="${value.asset.url}"></img>`;
      },
      component: ({value}) => {
        const linkedItem = dummyRichText.linkedItems.find(item => item.system.codename === value.component._ref);
        switch(linkedItem?.system.type) {
          case('test'): {
            return `<p>resolved value of text_element: <strong>${linkedItem?.elements.text_element.value}</strong></p>`;
          }
          default: {
            return `Resolver for type ${linkedItem?.system.type} not implemented.`
          }
        }
      },
      table: ({value}) => {
        const tableHtml = resolveTable(value, toHTML);
        return tableHtml;
      }
    },
    marks: {
      internalLink: ({children, value}) => {
        return `\<a href=\"https://website.com/${value.reference._ref}">${children}</a>`
      },
      link: ({ children, value }) => {
        return `\<a href=${escapeHTML(value.href)}">${children}</a>`
      }
    }
  }
}

describe("portable text transformer", () => {
  it("transforms empty rich text", () => {
    const tree = nodeParse(dummyRichText.value);
    const result = transform(tree);

    expect(result).toMatchSnapshot();
  })

  it("transforms tables", () => {
    dummyRichText.value = `<table><tbody>\n  <tr><td><p>paragraph 1</p><p>paragraph 2</p></td><td><ul>\n  <li>bulit</li>\n</ul>\n</td><td><a href=\"http://google.com\" data-new-window=\"true\" title=\"linktitle\" target=\"_blank\" rel=\"noopener noreferrer\">this is a<strong>strong</strong>link</a></td></tr>\n<tr><td><h1><strong>nadpis</strong></h1></td><td><p>text</p></td><td><p>text</p></td></tr>\n<tr><td><p>text</p></td><td><p>text</p></td><td><p>text</p></td></tr>\n</tbody></table>`
    const tree = nodeParse(dummyRichText.value);
    const result = transform(tree);

    expect(result).toMatchSnapshot();
  })

  it("transforms item links", () => {
    dummyRichText.value = `<p><a data-item-id=\"23f71096-fa89-4f59-a3f9-970e970944ec\" href=\"\">text<strong>link to an item</strong></a></p>`
    const tree = nodeParse(dummyRichText.value);
    const result = transform(tree);

    expect(result).toMatchSnapshot();
 
  })

  it("transforms external links", () => {
    dummyRichText.value = `<h2><strong>Kontent supports portable text!</strong></h2>\n<p>For more information, check out the related <a href=\"https://github.com/portabletext/portabletext\" data-new-window=\"true\" title=\"Portable text repo\" target=\"_blank\" rel=\"noopener noreferrer\"><strong>GitHub repository.</strong></a></p>`
    const tree = nodeParse(dummyRichText.value);
    const result = transform(tree);

    expect(result).toMatchSnapshot();
  })

  it("transforms nested styles", () => {
    dummyRichText.value = `<p><strong>all text is bold and last part is </strong><em><strong>also italic and this is also </strong></em><em><strong><sup>superscript</sup></strong></em></p>`
    const tree = nodeParse(dummyRichText.value);
    const result = transform(tree);

    expect(result).toMatchSnapshot();
  })

  it("transforms lists", () => {
    dummyRichText.value = `<ul><li>first level bullet</li><li>first level bullet</li><ol><li>nested number in bullet list</li></ol></ul><ol><li>first level item</li><li>first level item</li><ol><li>second level item</li><li><strong>second level item </strong></li></ol>`;
    const tree = nodeParse(dummyRichText.value);
    const result = transform(tree);

    expect(result).toMatchSnapshot();
  })

  it("transforms images", () => {
    dummyRichText.value = `<figure data-asset-id=\"7d866175-d3db-4a02-b0eb-891fb06b6ab0\" data-image-id=\"7d866175-d3db-4a02-b0eb-891fb06b6ab0\"><img src=\"https://assets-eu-01.kc-usercontent.com:443/6d864951-9d19-0138-e14d-98ba886a4410/236ecb7f-41e3-40c7-b0db-ea9c2c44003b/sharad-bhat-62p19OGT2qg-unsplash.jpg\" data-asset-id=\"7d866175-d3db-4a02-b0eb-891fb06b6ab0\" data-image-id=\"7d866175-d3db-4a02-b0eb-891fb06b6ab0\" alt=\"\"></figure><p><em>text in a paragraph</em></p>`;
    const tree = nodeParse(dummyRichText.value);
    const result = transform(tree);

    expect(result).toMatchSnapshot();
  })

  it("transforms complex rich text into portable text", () => {
    dummyRichText.value = "<table><tbody><tr><td><ul><li>list item</li><ol><li>nested list item</li></ol></ul></td></tr></tbody></table><table><tbody>\n  <tr><td><p>paragraph 1</p><p>paragraph 2</p></td><td><ul>\n  <li>list item\n     </li>\n</ul>\n</td><td><a href=\"http://google.com\" data-new-window=\"true\" title=\"linktitle\" target=\"_blank\" rel=\"noopener noreferrer\">this is a<strong>strong</strong>link</a></td></tr>\n<tr><td><h1><strong>nadpis</strong></h1></td><td><p>text</p></td><td><p>text</p></td></tr>\n<tr><td><em>italic text</em></td><td><p>text</p></td><td><p>text</p></td></tr>\n</tbody></table><p>text<a href=\"http://google.com\" data-new-window=\"true\" title=\"linktitle\" target=\"_blank\" rel=\"noopener noreferrer\">normal and<strong>bold</strong>link</a></p><h1>heading</h1><object type=\"application/kenticocloud\" data-type=\"item\" data-rel=\"link\" data-codename=\"test_item\"></object>";
    const tree = nodeParse(dummyRichText.value);
    const result = transform(tree);

    expect(result).toMatchSnapshot();
  })

  it("transforms linked items/components", () => {
    dummyRichText.value = "<object type=\"application/kenticocloud\" data-type=\"item\" data-rel=\"link\" data-codename=\"test_item\"></object>";
    const tree = nodeParse(dummyRichText.value);
    const result = transform(tree);

    expect(result).toMatchSnapshot();
  })

  it("ignores unknown html tags", () => {
    dummyRichText.value = "<p>text in a paragraph</p><div>text in a div, which doesnt exist in kontent RTE</div>"
    const tree = nodeParse(dummyRichText.value);
    const result = transform(tree);

    expect(result).toMatchSnapshot();
  })
})

describe("HTML converter", () => {
  it("builds basic portable text into HTML", () => {
    dummyRichText.value = '<p><br></p><p>text<a href=\"http://google.com\" data-new-window=\"true\" title=\"linktitle\" target=\"_blank\" rel=\"noopener noreferrer\"><strong>link</strong></a></p><h1>heading</h1><p><br></p>';
    const tree = nodeParse(dummyRichText.value);
    const portableText = transform(tree);
    const result = toHTML(portableText, portableTextComponents);

    expect(result).toMatchSnapshot();
  })

  it("resolves internal link", () => {
    dummyRichText.value = `<p><a data-item-id=\"23f71096-fa89-4f59-a3f9-970e970944ec\" href=\"\"><em>item</em></a></p>`
    const tree = nodeParse(dummyRichText.value);
    const portableText = transform(tree);
    const result = toHTML(portableText, portableTextComponents);

    expect(result).toMatchSnapshot();
  })

  it("resolves a linked item", () => {
    dummyRichText.value = '<object type=\"application/kenticocloud\" data-type=\"item\" data-rel=\"link\" data-codename=\"test_item\"></object><p>text after component</p>'
    const tree = nodeParse(dummyRichText.value);
    const portableText = transform(tree);
    const result = toHTML(portableText, portableTextComponents);
    
    expect(result).toMatchSnapshot();
  })

  it("resolves a table", () => {
    dummyRichText.value = "<table><tbody>\n  <tr><td>Ivan</td><td>Jiri</td></tr>\n  <tr><td>Ondra</td><td>Dan</td></tr>\n</tbody></table>";
    const jsonTree = nodeParse(dummyRichText.value);
    const portableText = transform(jsonTree);
    const result = toHTML(portableText, portableTextComponents);

    expect(result).toMatchSnapshot();
  })

  it("resolves an asset", () => {
    dummyRichText.value = `<figure data-asset-id=\"62ba1f17-13e9-43c0-9530-6b44e38097fc\" data-image-id=\"62ba1f17-13e9-43c0-9530-6b44e38097fc\"><img src=\"https://assets-us-01.kc-usercontent.com:443/cec32064-07dd-00ff-2101-5bde13c9e30c/3594632c-d9bb-4197-b7da-2698b0dab409/Riesachsee_Dia_1_1963_%C3%96sterreich_16k_3063.jpg\" data-asset-id=\"62ba1f17-13e9-43c0-9530-6b44e38097fc\" data-image-id=\"62ba1f17-13e9-43c0-9530-6b44e38097fc\" alt=\"\"></figure>`;
    const tree = nodeParse(dummyRichText.value);
    const portableText = transform(tree);
    const result = toHTML(portableText, portableTextComponents)

    expect(result).toMatchSnapshot();
  })
})