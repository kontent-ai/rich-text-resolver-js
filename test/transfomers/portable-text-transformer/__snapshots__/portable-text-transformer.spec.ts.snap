// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`HTML converter builds basic portable text into HTML 1`] = `"<p><br/></p><p>text<a href=http://google.com"><strong>link</strong></a></p><h1>heading</h1><p><br/></p>"`;

exports[`HTML converter resolves a linked item 1`] = `"<p>resolved value of text_element: <strong>random text value</strong></p><p>text after component</p>"`;

exports[`HTML converter resolves a table 1`] = `"<table><tbody><tr><td><p>Ivan</p></td><td><p>Jiri</p></td></tr><tr><td><p>Ondra</p></td><td><p>Dan</p></td></tr></tbody></table>"`;

exports[`HTML converter resolves an asset 1`] = `"<img src="https://assets-us-01.kc-usercontent.com:443/cec32064-07dd-00ff-2101-5bde13c9e30c/3594632c-d9bb-4197-b7da-2698b0dab409/Riesachsee_Dia_1_1963_%C3%96sterreich_16k_3063.jpg"></img>"`;

exports[`HTML converter resolves internal link 1`] = `"<p><a href="https://website.com/23f71096-fa89-4f59-a3f9-970e970944ec"><em>item</em></a></p>"`;

exports[`portable text transformer doesn't create duplicates for nested spans 1`] = `
[
  {
    "_key": "guid",
    "_type": "block",
    "children": [
      {
        "_key": "guid",
        "_type": "span",
        "marks": [],
        "text": "text",
      },
      {
        "_key": "guid",
        "_type": "span",
        "marks": [
          "strong",
        ],
        "text": "bold",
      },
    ],
    "markDefs": [],
    "style": "normal",
  },
]
`;

exports[`portable text transformer doesn't extend link mark to adjacent spans 1`] = `
[
  {
    "_key": "guid",
    "_type": "block",
    "children": [
      {
        "_key": "guid",
        "_type": "span",
        "marks": [],
        "text": "The coffee drinking culture has been evolving for hundreds and thousands of years. It has never been ",
      },
      {
        "_key": "guid",
        "_type": "span",
        "marks": [
          "guid",
        ],
        "text": "so rich as ",
      },
      {
        "_key": "guid",
        "_type": "span",
        "marks": [
          "strong",
          "guid",
        ],
        "text": "today",
      },
      {
        "_key": "guid",
        "_type": "span",
        "marks": [],
        "text": ". How do you make sense of different types of coffee, what do the particular names in beverage menus mean and what beverage to choose for which occasion in your favorite café?",
      },
    ],
    "markDefs": [
      {
        "_key": "guid",
        "_type": "internalLink",
        "reference": {
          "_ref": "3120ec15-a4a2-47ec-8ccd-c85ac8ac5ba5",
          "_type": "reference",
        },
      },
    ],
    "style": "normal",
  },
]
`;

exports[`portable text transformer flattens html to block array 1`] = `
[
  {
    "_key": "guid",
    "_type": "block",
    "children": [],
    "markDefs": [],
    "style": "normal",
  },
  {
    "_key": "guid",
    "_type": "span",
    "marks": [],
    "text": "The coffee drinking culture has been evolving for hundreds and thousands of years. It has never been ",
  },
  {
    "_key": "guid",
    "_type": "internalLink",
    "reference": {
      "_ref": "3120ec15-a4a2-47ec-8ccd-c85ac8ac5ba5",
      "_type": "reference",
    },
  },
  {
    "_key": "guid",
    "_type": "linkMark",
    "childCount": 2,
    "value": "guid",
  },
  {
    "_key": "guid",
    "_type": "span",
    "marks": [],
    "text": "so rich as ",
  },
  {
    "_key": "guid",
    "_type": "mark",
    "childCount": 1,
    "value": "strong",
  },
  {
    "_key": "guid",
    "_type": "span",
    "marks": [],
    "text": "today",
  },
  {
    "_key": "guid",
    "_type": "span",
    "marks": [],
    "text": ". How do you make sense of different types of coffee, what do the particular names in beverage menus mean and what beverage to choose for which occasion in your favorite café?",
  },
]
`;

exports[`portable text transformer transforms complex rich text into portable text 1`] = `
[
  {
    "_key": "guid",
    "_type": "table",
    "numColumns": 1,
    "rows": [
      {
        "_key": "guid",
        "_type": "row",
        "cells": [
          {
            "_key": "guid",
            "_type": "cell",
            "childBlocksCount": 2,
            "content": [
              {
                "_key": "guid",
                "_type": "block",
                "children": [
                  {
                    "_key": "guid",
                    "_type": "span",
                    "marks": [],
                    "text": "list item",
                  },
                ],
                "level": 1,
                "listItem": "bullet",
                "markDefs": [],
                "style": "normal",
              },
              {
                "_key": "guid",
                "_type": "block",
                "children": [
                  {
                    "_key": "guid",
                    "_type": "span",
                    "marks": [],
                    "text": "nested list item",
                  },
                ],
                "level": 2,
                "listItem": "number",
                "markDefs": [],
                "style": "normal",
              },
            ],
          },
        ],
      },
    ],
  },
  {
    "_key": "guid",
    "_type": "table",
    "numColumns": 3,
    "rows": [
      {
        "_key": "guid",
        "_type": "row",
        "cells": [
          {
            "_key": "guid",
            "_type": "cell",
            "childBlocksCount": 2,
            "content": [
              {
                "_key": "guid",
                "_type": "block",
                "children": [
                  {
                    "_key": "guid",
                    "_type": "span",
                    "marks": [],
                    "text": "paragraph 1",
                  },
                ],
                "markDefs": [],
                "style": "normal",
              },
              {
                "_key": "guid",
                "_type": "block",
                "children": [
                  {
                    "_key": "guid",
                    "_type": "span",
                    "marks": [],
                    "text": "paragraph 2",
                  },
                ],
                "markDefs": [],
                "style": "normal",
              },
            ],
          },
          {
            "_key": "guid",
            "_type": "cell",
            "childBlocksCount": 1,
            "content": [
              {
                "_key": "guid",
                "_type": "block",
                "children": [
                  {
                    "_key": "guid",
                    "_type": "span",
                    "marks": [],
                    "text": "list item",
                  },
                ],
                "level": 1,
                "listItem": "bullet",
                "markDefs": [],
                "style": "normal",
              },
            ],
          },
          {
            "_key": "guid",
            "_type": "cell",
            "childBlocksCount": 1,
            "content": [
              {
                "_key": "guid",
                "_type": "block",
                "children": [
                  {
                    "_key": "guid",
                    "_type": "span",
                    "marks": [
                      "guid",
                    ],
                    "text": "this is a",
                  },
                  {
                    "_key": "guid",
                    "_type": "span",
                    "marks": [
                      "strong",
                      "guid",
                    ],
                    "text": "strong",
                  },
                  {
                    "_key": "guid",
                    "_type": "span",
                    "marks": [
                      "guid",
                    ],
                    "text": "link",
                  },
                ],
                "markDefs": [
                  {
                    "_key": "guid",
                    "_type": "link",
                    "data-new-window": "true",
                    "href": "http://google.com",
                    "rel": "noopener noreferrer",
                    "target": "_blank",
                    "title": "linktitle",
                  },
                ],
                "style": "normal",
              },
            ],
          },
        ],
      },
      {
        "_key": "guid",
        "_type": "row",
        "cells": [
          {
            "_key": "guid",
            "_type": "cell",
            "childBlocksCount": 1,
            "content": [
              {
                "_key": "guid",
                "_type": "block",
                "children": [
                  {
                    "_key": "guid",
                    "_type": "span",
                    "marks": [
                      "strong",
                    ],
                    "text": "nadpis",
                  },
                ],
                "markDefs": [],
                "style": "h1",
              },
            ],
          },
          {
            "_key": "guid",
            "_type": "cell",
            "childBlocksCount": 1,
            "content": [
              {
                "_key": "guid",
                "_type": "block",
                "children": [
                  {
                    "_key": "guid",
                    "_type": "span",
                    "marks": [],
                    "text": "text",
                  },
                ],
                "markDefs": [],
                "style": "normal",
              },
            ],
          },
          {
            "_key": "guid",
            "_type": "cell",
            "childBlocksCount": 1,
            "content": [
              {
                "_key": "guid",
                "_type": "block",
                "children": [
                  {
                    "_key": "guid",
                    "_type": "span",
                    "marks": [],
                    "text": "text",
                  },
                ],
                "markDefs": [],
                "style": "normal",
              },
            ],
          },
        ],
      },
      {
        "_key": "guid",
        "_type": "row",
        "cells": [
          {
            "_key": "guid",
            "_type": "cell",
            "childBlocksCount": 1,
            "content": [
              {
                "_key": "guid",
                "_type": "block",
                "children": [
                  {
                    "_key": "guid",
                    "_type": "span",
                    "marks": [
                      "em",
                    ],
                    "text": "italic text",
                  },
                ],
                "markDefs": [],
                "style": "normal",
              },
            ],
          },
          {
            "_key": "guid",
            "_type": "cell",
            "childBlocksCount": 1,
            "content": [
              {
                "_key": "guid",
                "_type": "block",
                "children": [
                  {
                    "_key": "guid",
                    "_type": "span",
                    "marks": [],
                    "text": "text",
                  },
                ],
                "markDefs": [],
                "style": "normal",
              },
            ],
          },
          {
            "_key": "guid",
            "_type": "cell",
            "childBlocksCount": 1,
            "content": [
              {
                "_key": "guid",
                "_type": "block",
                "children": [
                  {
                    "_key": "guid",
                    "_type": "span",
                    "marks": [],
                    "text": "text",
                  },
                ],
                "markDefs": [],
                "style": "normal",
              },
            ],
          },
        ],
      },
    ],
  },
  {
    "_key": "guid",
    "_type": "block",
    "children": [
      {
        "_key": "guid",
        "_type": "span",
        "marks": [],
        "text": "text",
      },
      {
        "_key": "guid",
        "_type": "span",
        "marks": [
          "guid",
        ],
        "text": "normal and",
      },
      {
        "_key": "guid",
        "_type": "span",
        "marks": [
          "strong",
          "guid",
        ],
        "text": "bold",
      },
      {
        "_key": "guid",
        "_type": "span",
        "marks": [
          "guid",
        ],
        "text": "link",
      },
    ],
    "markDefs": [
      {
        "_key": "guid",
        "_type": "link",
        "data-new-window": "true",
        "href": "http://google.com",
        "rel": "noopener noreferrer",
        "target": "_blank",
        "title": "linktitle",
      },
    ],
    "style": "normal",
  },
  {
    "_key": "guid",
    "_type": "block",
    "children": [
      {
        "_key": "guid",
        "_type": "span",
        "marks": [],
        "text": "heading",
      },
    ],
    "markDefs": [],
    "style": "h1",
  },
  {
    "_key": "guid",
    "_type": "component",
    "component": {
      "_ref": "test_item",
      "_type": "reference",
    },
  },
]
`;

exports[`portable text transformer transforms empty rich text 1`] = `
[
  {
    "_key": "guid",
    "_type": "block",
    "children": [
      {
        "_key": "guid",
        "_type": "span",
        "marks": [],
        "text": "
",
      },
    ],
    "markDefs": [],
    "style": "normal",
  },
]
`;

exports[`portable text transformer transforms external links 1`] = `
[
  {
    "_key": "guid",
    "_type": "block",
    "children": [
      {
        "_key": "guid",
        "_type": "span",
        "marks": [
          "strong",
        ],
        "text": "Kontent supports portable text!",
      },
    ],
    "markDefs": [],
    "style": "h2",
  },
  {
    "_key": "guid",
    "_type": "block",
    "children": [
      {
        "_key": "guid",
        "_type": "span",
        "marks": [],
        "text": "For more information, check out the related ",
      },
      {
        "_key": "guid",
        "_type": "span",
        "marks": [
          "strong",
          "guid",
        ],
        "text": "GitHub repository.",
      },
    ],
    "markDefs": [
      {
        "_key": "guid",
        "_type": "link",
        "data-new-window": "true",
        "href": "https://github.com/portabletext/portabletext",
        "rel": "noopener noreferrer",
        "target": "_blank",
        "title": "Portable text repo",
      },
    ],
    "style": "normal",
  },
]
`;

exports[`portable text transformer transforms images 1`] = `
[
  {
    "_key": "guid",
    "_type": "image",
    "asset": {
      "_ref": "7d866175-d3db-4a02-b0eb-891fb06b6ab0",
      "_type": "reference",
      "url": "https://assets-eu-01.kc-usercontent.com:443/6d864951-9d19-0138-e14d-98ba886a4410/236ecb7f-41e3-40c7-b0db-ea9c2c44003b/sharad-bhat-62p19OGT2qg-unsplash.jpg",
    },
  },
  {
    "_key": "guid",
    "_type": "block",
    "children": [
      {
        "_key": "guid",
        "_type": "span",
        "marks": [
          "em",
        ],
        "text": "text in a paragraph",
      },
    ],
    "markDefs": [],
    "style": "normal",
  },
]
`;

exports[`portable text transformer transforms item links 1`] = `
[
  {
    "_key": "guid",
    "_type": "block",
    "children": [
      {
        "_key": "guid",
        "_type": "span",
        "marks": [
          "guid",
        ],
        "text": "text",
      },
      {
        "_key": "guid",
        "_type": "span",
        "marks": [
          "strong",
          "guid",
        ],
        "text": "link to an item",
      },
    ],
    "markDefs": [
      {
        "_key": "guid",
        "_type": "internalLink",
        "reference": {
          "_ref": "23f71096-fa89-4f59-a3f9-970e970944ec",
          "_type": "reference",
        },
      },
    ],
    "style": "normal",
  },
]
`;

exports[`portable text transformer transforms linked items/components 1`] = `
[
  {
    "_key": "guid",
    "_type": "component",
    "component": {
      "_ref": "test_item",
      "_type": "reference",
    },
  },
]
`;

exports[`portable text transformer transforms lists 1`] = `
[
  {
    "_key": "guid",
    "_type": "block",
    "children": [
      {
        "_key": "guid",
        "_type": "span",
        "marks": [],
        "text": "first level bullet",
      },
    ],
    "level": 1,
    "listItem": "bullet",
    "markDefs": [],
    "style": "normal",
  },
  {
    "_key": "guid",
    "_type": "block",
    "children": [
      {
        "_key": "guid",
        "_type": "span",
        "marks": [],
        "text": "first level bullet",
      },
    ],
    "level": 1,
    "listItem": "bullet",
    "markDefs": [],
    "style": "normal",
  },
  {
    "_key": "guid",
    "_type": "block",
    "children": [
      {
        "_key": "guid",
        "_type": "span",
        "marks": [],
        "text": "nested number in bullet list",
      },
    ],
    "level": 2,
    "listItem": "number",
    "markDefs": [],
    "style": "normal",
  },
  {
    "_key": "guid",
    "_type": "block",
    "children": [
      {
        "_key": "guid",
        "_type": "span",
        "marks": [],
        "text": "first level item",
      },
    ],
    "level": 1,
    "listItem": "number",
    "markDefs": [],
    "style": "normal",
  },
  {
    "_key": "guid",
    "_type": "block",
    "children": [
      {
        "_key": "guid",
        "_type": "span",
        "marks": [],
        "text": "first level item",
      },
    ],
    "level": 1,
    "listItem": "number",
    "markDefs": [],
    "style": "normal",
  },
  {
    "_key": "guid",
    "_type": "block",
    "children": [
      {
        "_key": "guid",
        "_type": "span",
        "marks": [],
        "text": "second level item",
      },
    ],
    "level": 2,
    "listItem": "number",
    "markDefs": [],
    "style": "normal",
  },
  {
    "_key": "guid",
    "_type": "block",
    "children": [
      {
        "_key": "guid",
        "_type": "span",
        "marks": [
          "strong",
        ],
        "text": "second level item ",
      },
    ],
    "level": 2,
    "listItem": "number",
    "markDefs": [],
    "style": "normal",
  },
]
`;

exports[`portable text transformer transforms nested styles 1`] = `
[
  {
    "_key": "guid",
    "_type": "block",
    "children": [
      {
        "_key": "guid",
        "_type": "span",
        "marks": [
          "strong",
        ],
        "text": "all text is bold and last part is ",
      },
      {
        "_key": "guid",
        "_type": "span",
        "marks": [
          "em",
          "strong",
        ],
        "text": "also italic and this is also ",
      },
      {
        "_key": "guid",
        "_type": "span",
        "marks": [
          "em",
          "strong",
          "sup",
        ],
        "text": "superscript",
      },
    ],
    "markDefs": [],
    "style": "normal",
  },
]
`;

exports[`portable text transformer transforms tables 1`] = `
[
  {
    "_key": "guid",
    "_type": "table",
    "numColumns": 3,
    "rows": [
      {
        "_key": "guid",
        "_type": "row",
        "cells": [
          {
            "_key": "guid",
            "_type": "cell",
            "childBlocksCount": 2,
            "content": [
              {
                "_key": "guid",
                "_type": "block",
                "children": [
                  {
                    "_key": "guid",
                    "_type": "span",
                    "marks": [],
                    "text": "paragraph 1",
                  },
                ],
                "markDefs": [],
                "style": "normal",
              },
              {
                "_key": "guid",
                "_type": "block",
                "children": [
                  {
                    "_key": "guid",
                    "_type": "span",
                    "marks": [],
                    "text": "paragraph 2",
                  },
                ],
                "markDefs": [],
                "style": "normal",
              },
            ],
          },
          {
            "_key": "guid",
            "_type": "cell",
            "childBlocksCount": 1,
            "content": [
              {
                "_key": "guid",
                "_type": "block",
                "children": [
                  {
                    "_key": "guid",
                    "_type": "span",
                    "marks": [],
                    "text": "bulit",
                  },
                ],
                "level": 1,
                "listItem": "bullet",
                "markDefs": [],
                "style": "normal",
              },
            ],
          },
          {
            "_key": "guid",
            "_type": "cell",
            "childBlocksCount": 1,
            "content": [
              {
                "_key": "guid",
                "_type": "block",
                "children": [
                  {
                    "_key": "guid",
                    "_type": "span",
                    "marks": [
                      "guid",
                    ],
                    "text": "this is a",
                  },
                  {
                    "_key": "guid",
                    "_type": "span",
                    "marks": [
                      "strong",
                      "guid",
                    ],
                    "text": "strong",
                  },
                  {
                    "_key": "guid",
                    "_type": "span",
                    "marks": [
                      "guid",
                    ],
                    "text": "link",
                  },
                ],
                "markDefs": [
                  {
                    "_key": "guid",
                    "_type": "link",
                    "data-new-window": "true",
                    "href": "http://google.com",
                    "rel": "noopener noreferrer",
                    "target": "_blank",
                    "title": "linktitle",
                  },
                ],
                "style": "normal",
              },
            ],
          },
        ],
      },
      {
        "_key": "guid",
        "_type": "row",
        "cells": [
          {
            "_key": "guid",
            "_type": "cell",
            "childBlocksCount": 1,
            "content": [
              {
                "_key": "guid",
                "_type": "block",
                "children": [
                  {
                    "_key": "guid",
                    "_type": "span",
                    "marks": [
                      "strong",
                    ],
                    "text": "nadpis",
                  },
                ],
                "markDefs": [],
                "style": "h1",
              },
            ],
          },
          {
            "_key": "guid",
            "_type": "cell",
            "childBlocksCount": 1,
            "content": [
              {
                "_key": "guid",
                "_type": "block",
                "children": [
                  {
                    "_key": "guid",
                    "_type": "span",
                    "marks": [],
                    "text": "text",
                  },
                ],
                "markDefs": [],
                "style": "normal",
              },
            ],
          },
          {
            "_key": "guid",
            "_type": "cell",
            "childBlocksCount": 1,
            "content": [
              {
                "_key": "guid",
                "_type": "block",
                "children": [
                  {
                    "_key": "guid",
                    "_type": "span",
                    "marks": [],
                    "text": "text",
                  },
                ],
                "markDefs": [],
                "style": "normal",
              },
            ],
          },
        ],
      },
      {
        "_key": "guid",
        "_type": "row",
        "cells": [
          {
            "_key": "guid",
            "_type": "cell",
            "childBlocksCount": 1,
            "content": [
              {
                "_key": "guid",
                "_type": "block",
                "children": [
                  {
                    "_key": "guid",
                    "_type": "span",
                    "marks": [],
                    "text": "text",
                  },
                ],
                "markDefs": [],
                "style": "normal",
              },
            ],
          },
          {
            "_key": "guid",
            "_type": "cell",
            "childBlocksCount": 1,
            "content": [
              {
                "_key": "guid",
                "_type": "block",
                "children": [
                  {
                    "_key": "guid",
                    "_type": "span",
                    "marks": [],
                    "text": "text",
                  },
                ],
                "markDefs": [],
                "style": "normal",
              },
            ],
          },
          {
            "_key": "guid",
            "_type": "cell",
            "childBlocksCount": 1,
            "content": [
              {
                "_key": "guid",
                "_type": "block",
                "children": [
                  {
                    "_key": "guid",
                    "_type": "span",
                    "marks": [],
                    "text": "text",
                  },
                ],
                "markDefs": [],
                "style": "normal",
              },
            ],
          },
        ],
      },
    ],
  },
]
`;
